product_multi_barcode_constraint_per_company/models/product_barcode.py

```diff
@@ -0,0 +1,35 @@
+# Copyright 2023 Cetmix OÜ
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, api, models
+from odoo.exceptions import UserError
+
+
+class ProductBarcode(models.Model):
+    _inherit = "product.barcode"
+
+    @api.constrains("name")
+    def _check_duplicates(self):
+        barcodes = self.search(
+            [("id", "not in", self.ids), ("name", "in", self.mapped("name"))]
+        )
+
+        for barcode in barcodes:
+            for record in self:
+                if (
+                    barcode
+                    and barcode.sudo().product_tmpl_id.company_id.id
+                    == record.product_tmpl_id.company_id.id
+                ):
+                    product = barcode.sudo().product_id
+                    raise UserError(
+                        _(
+                            'The Barcode "%(barcode_name)s" already exists for '
+                            'product "%(product_name)s" in the company %(company_name)s'
+                        )
+                        % dict(
```

Comment:
`dict()` slowly than `{}`, so better use `{}` in code.

Reply:
okay, thanks